kratos/python_scripts/restart_utility.py

```diff
@@ -170,6 +191,38 @@ def CreateOutputFolder(self):
                 os.makedirs(folder_path)
             self.model_part.GetCommunicator().GetDataCommunicator().Barrier()
 
+    def GetRestartFiles(self):
+        """Return a dictionary of stepID - restart_file_list dictionary that stores sets of restart files for each step."""
+        restart_path    = Path( self.__GetFolderPathSave() )
+        restart_files   = {}
+        if restart_path.is_dir():
+            for entry in restart_path.glob("**/*"):
+                if entry.is_file() and self.__IsRestartFile( entry.name ):
+                    label = self.__ExtractFileLabel(entry.name) # Get process ID and step ID
+                    if label[1] in restart_files:               # Check if this step has entries already
+                        restart_files[label[1]].append( entry.name )
+                    else:
+                        restart_files[label[1]] = [ entry.name ]
+
+        return restart_files
+
+    def ClearObsoleteRestartFiles(self):
+        """Delete restart files that are no longer needed."""
+        if self.number_of_restart_files > -1:               # <-- number of restart files is limited
+            number_of_obsolete_files = len(self.restart_files) - self.number_of_restart_files
+            for _ in range(number_of_obsolete_files):
+
+                # Get oldest restart file set
+                oldest_step_id, oldest_file_set = min( self.restart_files.items(), key=lambda item: float(item[0]) )
```

Comment:
OpenMP parallelism works, with MPI my test fails with this error:

```
  File "/dss/dsshome1/0E/di49xub3/software/KratosTest/bin/Release/KratosMultiphysics/restart_utility.py", line 216, in <lambda>
    oldest_step_id, oldest_file_set = min( self.restart_files.items(), key=lambda item: float(item[0]) )
ValueError: could not convert string to float: '0_0.4'
```

The files are labeled as follows:
```
-rw-r--r-- 1 di49xub3 hpcusers 44497140 23. Sep 13:00 FluidModelPart_0_0.16.rest
-rw-r--r-- 1 di49xub3 hpcusers 44497140 23. Sep 13:01 FluidModelPart_0_0.24.rest
```

Reply:
Looks like I forgot to split file labels when generating new restart files. I think this commit fixes it, please run your test again.

Reply:
It ran on MPI. I did not re-test with OpenMP.